gunrock/util/kernel_segmode.hxx

```diff
@@ -0,0 +1,1259 @@
+
+          merge = mgpu::segmented_serial_merge<vt>(shared.keys, 
+            local_range.partition(mp, diag), active, comp, false);
+        } else {
+         mgpu::iterate<vt>([&](int i) {
+            merge.indices[i] = vt * tid + i;
+          });
+        }
+        __syncthreads();
+
+        // Store keys to global memory.
+        if(sort_warp)
+          mgpu::reg_to_shared_thread<nt, vt>(merge.keys, tid, shared.keys, false);
+        __syncthreads();
+
+        mgpu::shared_to_mem<nt, vt>(shared.keys, tid, count2, keys_dest + first);
+
+        // if(has_values) {
+        //   // Transpose the indices from thread order to strided order.
+        //   array_t<int, vt> indices = reg_thread_to_strided<nt>(merge.indices,
+        //     tid, shared.indices);
+
+        //   // Gather the input values and merge into the output values.
+        //   transfer_two_streams_strided<nt>(vals_source + range.a_begin, 
+        //     range.a_count(), vals_source + range.b_begin, range.b_count(), 
+        //     indices, tid, vals_dest + first);
+        // }
+                
+
+      };
+      mgpu::cta_launch<launch_t>(merge_k, &op_counters_data[pass].x, context);
+        // printf("%d, \n", head_flags_saved_data[tid]);
+        // printf("%d,%d \n", &op_counters_data[pass].x, &op_counters_data[pass].y);
+
+      auto copy_k = [=] MGPU_DEVICE(int tid, int cta) {
+        typedef typename launch_t::sm_ptx params_t;
+        enum { nt = params_t::nt, vt = params_t::vt, nv = nt * vt };
+
+        int tile = copy_list_data[cta];
+        int first = nv * tile;
+        int count2 = min((int)nv, count - first);
+
+        mgpu::mem_to_mem<nt, vt>(keys_source + first, tid, count2, 
+          keys_dest + first);
+
+        // if(has_values)
+        //   mgpu::mem_to_mem<nt, vt>(vals_source + first, tid, count2, 
+        //     vals_dest + first);
+      };
+      mgpu::cta_launch<launch_t>(copy_k, &op_counters_data[pass].y, context);
+
+      std::swap(keys_source, keys_dest);
+      std::swap(vals_source, vals_dest);
+    }    
+    }
+  }
+};
+
+} // namespace detail
+
+
+template<typename comp_t>
+int* segmented_mode(int* keys, int count, int* segments, int num_segments, comp_t comp, mgpu::context_t& context) {
+  
+
+
+  detail::segsort_t<mgpu::empty_t, int, int, comp_t> 
+    segsort(keys,  (int*)nullptr, count, comp, context);
+  segsort.template blocksort_segments<true>(keys, mgpu::counting_iterator_t<int>(), 
+    segments, num_segments);
+  segsort.merge_passes();
+  mgpu::mem_t<quad> results(num_segments, context);
+  quad init = {-1,0,-1,0,-1,0,-1};
+
+  segreduce<mgpu::empty_t>(keys, count, segments, 
+  num_segments, results.data(), perform_t<quad>(), init, context);
+  std::vector<quad> results_host = from_mem(results);
+  int* answer_host = (int*)malloc(sizeof(int)*num_segments);
+
+  for(int i = 0;i<num_segments;i++){
+    answer_host[i] = results_host[i].best_element;
+  }
+
+  return answer_host;
+}
```

Comment:
Is this the final function that is exposed to the user? 

If so, I'd really like to see some tests for segmented_mode. There are many smaller helper functions that seem to contribute to this final utility function. Adding several tests will make it easier for me to review this large file.